### PR TITLE
DmChar08 OK

### PR DESCRIPTION
**Pull request by hensldm**
Supersedes #790 thank you @louist103 for the groundwork done.

### CODE REVIEW

File: assets/xml/objects/object_kamejima.xml [hunk 1]
```diff
@@ -1,81 +1,99 @@
...
+        <TextureAnimation Name="gTurtleWaveTexAnim" Offset="0x1AF0" />
+
+
+        <!-- Turtle -->
+        <Collision Name="gTurtleAsleepCol" Offset="0x2328" />
+        <TextureAnimation Name="gTurtleEmptyTexAnim" Offset="0x2360"/>
+        <Collision Name="gTurtleAwakeCol" Offset="0x2470" />
+        <Collision Name="gTurtleUnusedCol" Offset="0x2E04" />
+        
+        <Animation Name="gTurtleHide1Anim" Offset="0x3980" />
```
Comment (tom-overton) [inline assets/xml/objects/object_kamejima.xml:20]: I wondered why there were so many "hide" animations, so I took a look into this object, and I think you're being bitten by some relatively undocumented changes in Z64Utils. In Random06457's original version of Z64Utils, trying to load these animations will throw an exception:
image

However, in both Skawo's fork and my fork, these animations will instead show the skeleton in its default "folded" state; it just so happens that in the Turtle's case, this kind of looks like he's hiding. I ran into a similar thing with `object_az`. Loading the animation in-game shows that it *does* do something, even if it's very glitchy:

https://user-images.githubusercontent.com/12245827/191217877-c111fc2f-497e-4639-a35d-9bc2a9f81d62.mp4

So in short, these probably aren't hiding animations, and you'll probably need to just to view them in-game to document them. I'll add a warning in my Z64Utils fork for when this "folded state override" occurs.

Comment (hensldm) [reply assets/xml/objects/object_kamejima.xml:20]: Ah I did not know that was the fallback behavior on errors. Thanks for the heads up. I did go and give them some different names.

**Comment by hensldm** [comment]
One other small note, I found a better match for init by just using `thisx`.

File: assets/xml/objects/object_kamejima.xml (hunk 1 above)
Comment (hensldm) [reply assets/xml/objects/object_kamejima.xml]: https://user-images.githubusercontent.com/31974442/191390104-ebc23015-e671-4279-b2d0-94a8e2c2441c.mp4

For reference the turtle is not having a very good day.

Comment (tom-overton) [reply assets/xml/objects/object_kamejima.xml]: While I knew they would probably be glitchy, I certainly didn't expect that, lol

**Review by Kenix3 — APPROVED** [review]

File: src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c [hunk 2]
```diff
@@ -28,71 +63,1061 @@ const ActorInit Dm_Char08_InitVars = {
...
+void DmChar08_Init(Actor* thisx, PlayState* play2) {
+    PlayState* play = play2;
+    DmChar08* this = THIS;
+
+    thisx->targetMode = 5;
+    this->eyeMode = TURTLE_EYEMODE_2;
+    thisx->targetArrowOffset = 120.0f;
+    ActorShape_Init(&thisx->shape, 0.0f, ActorShadow_DrawCircle, 24.0f);
+    SkelAnime_InitFlex(play, &this->skelAnime, &gTurtleSkel, NULL, NULL, NULL, 0);
+    Actor_ProcessInitChain(&this->dyna.actor, sInitChain);
```
Comment (tom-overton) [inline src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c:155]: Are you required to use `&this->dyna.actor` here, or can you use `thisx` like other parts of this function?
```suggestion
    Actor_ProcessInitChain(thisx, sInitChain);
```

File: src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c [hunk 3]
```diff
@@ -28,71 +63,1061 @@ const ActorInit Dm_Char08_InitVars = {
...
+
+void DmChar08_Destroy(Actor* thisx, PlayState* play) {
+    DmChar08* this = THIS;
+
+    if (this->dynapolyInitialized) {
+        DynaPoly_DeleteBgActor(play, &play->colCtx.dyna, this->dyna.bgId);
+    }
+}
+
+s16 sBigTurtleOcarinaSoundLatch = false;
```
Comment (tom-overton) [inline src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c:251]: If I remember correctly, there was some debate around this name in the original PR. I suggested `sSuccessSoundAlreadyPlayed` at the time; what do you think?

File: src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c [hunk 4]
```diff
@@ -28,71 +63,1061 @@ const ActorInit Dm_Char08_InitVars = {
...
+        this->actionFunc = func_80AAFB94;
+    }
+}
+
+void func_80AAFB94(DmChar08* this, PlayState* play) {
+}
+
+Color_RGBA8 sLargeTurtleBubblePrimColor = { 250, 200, 250, 100 };
+Color_RGBA8 sLargeTurtleBubbleEnvColor = { 80, 80, 80, 0 };
+Vec3f sLargeTurtleBubbelAccel = { 0.0f, -0.8f, 0.0f };
```
Comment (tom-overton) [inline src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c:371]:
```suggestion
Vec3f sLargeTurtleBubbleAccel = { 0.0f, -0.8f, 0.0f };
```

File: src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c [hunk 5]
```diff
@@ -28,71 +63,1061 @@ const ActorInit Dm_Char08_InitVars = {
...
+            this->eyeIndex = 2;
+            break;
+
+        case TURTLE_EYEMODE_3:
+            this->eyeIndex = 4;
+            break;
+
+        case TURTLE_EYEMODE_5:
+            this->eyeIndex = 8;
+            break;
```
Comment (tom-overton) [inline src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c:126]: Maybe you can't name all of them, but some of these can be named, right? Since they just set a static `eyeIndex`.

File: assets/xml/objects/object_kamejima.xml [hunk 6]
```diff
@@ -1,81 +1,99 @@
...
+        <DList Name="gTurtleWaveDL" Offset="0x140" />
+        <Texture Name="gTurtleWaveUnusedTex" OutName="turtle_wave_unused" Format="i8" Width="32" Height="64" Offset="0x2D8" />
+        <Texture Name="gTurtleWaveTex" OutName="turtle_wave" Format="i8" Width="64" Height="64" Offset="0xAD8" />
+        <TextureAnimation Name="gTurtleWaveTexAnim" Offset="0x1AF0" />
+
+
+        <!-- Turtle -->
+        <Collision Name="gTurtleAsleepCol" Offset="0x2328" />
+        <TextureAnimation Name="gTurtleEmptyTexAnim" Offset="0x2360"/>
+        <Collision Name="gTurtleAwakeCol" Offset="0x2470" />
```
[the remark this answers is not in the document]
Comment (hensldm) [reply assets/xml/objects/object_kamejima.xml:17]: I wouldn't be against. I was mainly going with specifying the Great Bay Temple variation, with having the Zora Cape version being the default (i.e. not special qualifier), though it is hard to see here since the Great Bay Temple assets are in the overlay itself. If qualifying the Zora Cape version as well is preferred just let me know.

File: assets/xml/objects/object_kamejima.xml [hunk 7]
```diff
@@ -1,81 +1,99 @@
...
+        <Collision Name="gTurtleUnusedCol" Offset="0x2E04" />
+        
+        <Animation Name="gTurtleGlitchy1Anim" Offset="0x3980" />
+        <Animation Name="gTurtleCoughAnim" Offset="0x47B8" />
+        <Animation Name="gTurtleIdleAnim" Offset="0x48B0" />
+
+        <DList Name="gTurtleAsleepDL" Offset="0x4E70" />
```
[the remark this answers is not in the document]
Comment (hensldm) [reply assets/xml/objects/object_kamejima.xml:24]: I get where you are coming from, though don't think it is necessary to lengthen the name since it still is the form the turtle takes when he is asleep. Perhaps a comment explaining that he takes the form of a rock when he's asleep would be enough?

File: src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c (hunk 5 above)
Comment (hensldm) [reply src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c]: Wasn't too bad just naming them all.

File: assets/xml/objects/object_kamejima.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (tom-overton) [reply assets/xml/objects/object_kamejima.xml:24]: Yeah, that would be totally fine with me. I just want something in the codebase that says "hey, he's a rock here, not a turtle" so that people who might be a little less familiar with the game don't get confused.

File: assets/xml/objects/object_kamejima.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (tom-overton) [reply assets/xml/objects/object_kamejima.xml:17]: It's a little more clear if you look at the code:
```c
    if (play->sceneNum == SCENE_31MISAKI) {
        if (gSaveContext.save.weekEventReg[53] & 0x20) {
            DynaPolyActor_Init(&this->dyna, 3);
            DynaPolyActor_LoadMesh(play, &this->dyna, &gTurtleAwakeCol);
        } else {
            DynaPolyActor_Init(&this->dyna, 3);
            DynaPolyActor_LoadMesh(play, &this->dyna, &gTurtleAsleepCol);
        }
        this->dynapolyInitialized = true;
    } else if (play->sceneNum == SCENE_SEA) {
        DynaPolyActor_Init(&this->dyna, 3);
        DynaPolyActor_LoadMesh(play, &this->dyna, &sTurtleGreatBayTempleCol);
        this->dynapolyInitialized = true;
    }
```
That being said, I think I'd prefer it if these ones specify that they're the Zora Cape versions. At the moment, the asset itself doesn't really explain that, and you have to look at the code to see how it's being used.

[the remark this answers is not in the document]
Comment (hensldm) [reply assets/xml/objects/object_kamejima.xml]: Added. Let me know if there is something else that deserved the same treatment.

File: src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c [hunk 8]
```diff
@@ -28,71 +63,1061 @@ const ActorInit Dm_Char08_InitVars = {
...
+
+                    case 1:
+                        // "Why so slow?" ...
+                        Message_StartTextbox(play, 0x102A, &this->dyna.actor);
+                        this->unk_206++;
+                        break;
+
+                    case 2:
+                        // "Heh, heh, heh.  Of course I know your name."...
+                        if (Message_GetState(&play->msgCtx) == 2) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c:406]:
```suggestion
                        if (Message_GetState(&play->msgCtx) == TEXT_STATE_CLOSING) {
```

File: src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c [hunk 9]
```diff
@@ -28,71 +63,1061 @@ const ActorInit Dm_Char08_InitVars = {
...
+                        break;
+
+                    case 1:
+                        //"...Are you going to the temple again?"...
+                        Message_StartTextbox(play, 0x103B, &this->dyna.actor);
+                        this->unk_206++;
+                        break;
+
+                    case 2:
+                        if (Message_GetState(&play->msgCtx) == 2) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c:424]:
```suggestion
                        if (Message_GetState(&play->msgCtx) == TEXT_STATE_CLOSING) {
```

File: src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c [hunk 10]
```diff
@@ -28,71 +63,1061 @@ const ActorInit Dm_Char08_InitVars = {
...
+                case 0:
+                    break;
+
+                case 1:
+                    Message_StartTextbox(play, 0x102E, &this->dyna.actor);
+                    this->unk_206++;
+                    break;
+
+                case 2:
+                    if (Message_GetState(&play->msgCtx) == 2) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c:444]:
```suggestion
                    if (Message_GetState(&play->msgCtx) == TEXT_STATE_CLOSING) {
```

File: src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c [hunk 11]
```diff
@@ -28,71 +63,1061 @@ const ActorInit Dm_Char08_InitVars = {
...
+                    }
+                    break;
+
+                case 2:
+                    if (Animation_OnFrame(&this->skelAnime, this->skelAnime.endFrame)) {
+                        this->animIndex = TURTLE_ANIM_FLOAT;
+                        this->eyeMode = TURTLE_EYEMODE_BLINK_LEFT;
+                        this->unk_208++;
+                    }
+                    if ((Message_GetState(&play->msgCtx) == 6) && Message_ShouldAdvance(play)) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c:838]:
```suggestion
                    if ((Message_GetState(&play->msgCtx) == TEXT_STATE_DONE) && Message_ShouldAdvance(play)) {
```

File: src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c [hunk 12]
```diff
@@ -28,71 +63,1061 @@ const ActorInit Dm_Char08_InitVars = {
...
+                    this->animIndex = TURTLE_ANIM_SPEAK1;
+                    this->eyeMode = TURTLE_EYEMODE_BLINK_LEFT;
+                    this->unk_208++;
+                    break;
+
+                case 1:
+                    if (Animation_OnFrame(&this->skelAnime, this->skelAnime.endFrame)) {
+                        this->unk_208++;
+                    }
+                    if ((Message_GetState(&play->msgCtx) == 6) && Message_ShouldAdvance(play)) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c:827]:
```suggestion
                    if ((Message_GetState(&play->msgCtx) == TEXT_STATE_DONE) && Message_ShouldAdvance(play)) {
```

File: src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c [hunk 13]
```diff
@@ -28,71 +63,1061 @@ const ActorInit Dm_Char08_InitVars = {
...
+    DmChar08* this = THIS;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Dm_Char08/func_80AB0F90.s")
+    OPEN_DISPS(play->state.gfxCtx);
+    func_8012C28C(play->state.gfxCtx);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Dm_Char08/DmChar08_Draw.s")
+    gSPSegment(POLY_OPA_DISP++, 0x08, Lib_SegmentedToVirtual(sBigTurtleEyeTextures[this->eyeIndex]));
+    gSPSegment(POLY_OPA_DISP++, 0x09, Lib_SegmentedToVirtual(sBigTurtleEyeTextures[this->eyeIndex]));
+    if ((this->unk_1FF > 0) || (play->csCtx.state != 0)) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c:1100]:
```suggestion
    if ((this->unk_1FF > 0) || (play->csCtx.state != CS_STATE_0)) {
```

**Review by AngheloAlf — APPROVED** [review]

File: src/overlays/actors/ovl_Dm_Char08/z_dm_char08.h [hunk 14]
```diff
@@ -8,11 +8,37 @@ struct DmChar08;
...
+    /* 0x201 */ u8 eyeIndex;
+    /* 0x202 */ u8 animIndex;
+    /* 0x203 */ u8 unk_203;
+    /* 0x204 */ UNK_TYPE1 unk_204;
+    /* 0x205 */ u8 alpha;
+    /* 0x206 */ u8 unk_206;
+    /* 0x207 */ u8 unk_207;
+    /* 0x208 */ u8 unk_208;
+    /* 0x209 */ u8 dynapolyInitialized;
+} DmChar08;// Size = 0x020C
```
Comment (engineer124) [inline src/overlays/actors/ovl_Dm_Char08/z_dm_char08.h:41]:
```suggestion
    /* 0x209 */ u8 dynapolyInitialized;
} DmChar08; // Size = 0x20C 
```

File: src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c [hunk 15]
```diff
@@ -28,71 +63,1061 @@ const ActorInit Dm_Char08_InitVars = {
...
+                this->actionFunc = DmChar08_SetupAppearCs;
+            }
+        }
+    }
+}
+
+void DmChar08_SetupAppearCs(DmChar08* this, PlayState* play) {
+    s16 cs1 = this->dyna.actor.cutscene;
+    s16 cs = ActorCutscene_GetAdditionalCutscene(
+        ActorCutscene_GetAdditionalCutscene(ActorCutscene_GetAdditionalCutscene(cs1)));
```
Comment (engineer124) [inline src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c:280]: Well this is something

File: src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c [hunk 16]
```diff
@@ -28,71 +63,1061 @@ const ActorInit Dm_Char08_InitVars = {
...
+                        break;
+                }
+            } else {
+                switch (this->unk_206) {
+                    case 0:
+                        break;
+
+                    case 1:
+                        //"...Are you going to the temple again?"...
+                        Message_StartTextbox(play, 0x103B, &this->dyna.actor);
```
Comment (engineer124) [inline src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c:419]: This looks like an exact quote. Could you summarize this?

File: src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c [hunk 17]
```diff
@@ -28,71 +63,1061 @@ const ActorInit Dm_Char08_InitVars = {
...
+}
+
+void func_80AB096C(DmChar08* this, PlayState* play) {
+    if ((play->csCtx.state != 0) && (play->sceneId == SCENE_31MISAKI) && (gSaveContext.sceneLayer == 0) &&
+        (play->csCtx.currentCsIndex == 0)) {
+        if ((play->csCtx.frames >= 890) && (play->csCtx.frames < 922)) {
+            Actor_PlaySfxAtPos(&this->dyna.actor, NA_SE_EV_EARTHQUAKE_LAST2 - SFX_FLAG);
+        }
+    }
+    if ((this->animIndex == 1) && Animation_OnFrame(&this->skelAnime, 16.0f)) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c:902]:
```suggestion
    if ((this->animIndex == TURTLE_ANIM_SWIM) && Animation_OnFrame(&this->skelAnime, 16.0f)) {
```

File: src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c [hunk 18]
```diff
@@ -28,71 +63,1061 @@ const ActorInit Dm_Char08_InitVars = {
...
+}
+
+void func_80AAFE88(DmChar08* this, PlayState* play) {
+    s32 actorActionIndex;
+    CsCmdActorAction* csAction;
+    f32 phi_f12;
+
+    if (Cutscene_CheckActorAction(play, 474)) {
+        actorActionIndex = Cutscene_GetActorActionIndex(play, 474);
+        if (this->unk_1F6 != play->csCtx.actorActions[actorActionIndex]->action) {
```
Comment (jpburnett) [inline src/overlays/actors/ovl_Dm_Char08/z_dm_char08.c:464]: Other places in the code base I've seen this called `csAction`, don't know if the same could apply here.

**Review by jpburnett — APPROVED** [review]
